src/compiler/c.rs

```diff
@@ -255,19 +261,21 @@ pub const CACHED_ENV_VARS : &'static [&'static str] = &[
 ];
 
 /// Compute the hash key of `compiler` compiling `preprocessor_output` with `args`.
-pub fn hash_key(compiler_digest: &str, arguments: &str, preprocessor_output: &[u8]) -> String {
+pub fn hash_key(compiler_digest: &str, arguments: &str, env_vars: &[(OsString, OsString)],
+                preprocessor_output: &[u8]) -> String
+{
     // If you change any of the inputs to the hash, you should change `CACHE_VERSION`.
     let mut m = sha1::Sha1::new();
     m.update(compiler_digest.as_bytes());
     m.update(CACHE_VERSION);
     m.update(arguments.as_bytes());
-    //TODO: should propogate these over from the client.
-    // https://github.com/glandium/sccache/issues/5
-    for var in CACHED_ENV_VARS.iter() {
-        if let Ok(val) = env::var(var) {
-            m.update(var.as_bytes());
+    //TODO: use lazy_static.
+    let cached_env_vars: HashSet<OsString> = CACHED_ENV_VARS.iter().map(|v| OsStr::new(v).to_os_string()).collect();
+    for &(ref var, ref val) in env_vars.iter() {
+        if cached_env_vars.contains(var) {
+            m.update(os_str_bytes(var));
```

Comment:
This runs a risk of breaking in future rust versions, but could `Hash for OsStr` be used here instead?